Userland/Applications/Dictionary/Dictionary.cpp

```diff
@@ -0,0 +1,183 @@
+    m_dictionary_bytes = m_dictionary_file->data();
+
+    const uint32_t* header = (const uint32_t*)m_dictionary_bytes;
+
+    m_num_words = header[0];
+    m_max_word_length = header[1];
+
+    m_index = (decltype(m_index))(
+        (const uint8_t*)m_dictionary_bytes + header[2]);
+
+    m_word_data_index = (decltype(m_word_data_index))(
+        (const uint8_t*)m_dictionary_bytes + header[3]);
+
+    printf("Number of words: %i\n", m_num_words);
+    printf("Max word length: %i\n", m_max_word_length);
+
+    printf("offsets:\n");
+    printf("  index: %u\n", header[2]);
+    printf("  wdata: %u\n", header[3]);
+    printf("  defin: %u\n", header[4]);
+
+    assert(m_num_words < 1000 * 1000);
+    assert(m_max_word_length == 16);
+}
+
+AK::StringView Dictionary::word_at_index(int word_index) const
+{
+    if (0 <= word_index && word_index < m_num_words) {
+        const char* c = (const char*)(m_index[word_index]);
+
+        int n = 0;
+        for (int i = 0; i < 16; i++) {
+            if (c[i])
+                n++;
+            else
+                break;
+        }
+
+        //printf("%i. %s\n", word_index, c);
+
+        return AK::StringView(c, n);
+    }
+    return "";
+}
+
+AK::String Dictionary::definition_of(int word_index) const
+{
+    if (!(0 <= word_index && word_index < m_num_words)) {
+        return "";
+    }
+
+    uint32_t word_data_offset = m_word_data_index[word_index];
+
+    WordData& word_data = *(WordData*)((uint8_t*)m_dictionary_bytes + word_data_offset);
+
+    AK::StringBuilder string_builder;
+
+    string_builder.append(word_at_index(word_index));
+    string_builder.append("\n\n");
+
+    char last_part_of_speech = 0;
+
+    int n = word_data.num_senses;
+    for (int i = 0; i < n; i++) {
+        Sense& sense = word_data.senses[i];
+
+        Definition& definition = *(Definition*)((uint8_t*)m_dictionary_bytes + sense.definition_offset);
+
+        if (definition.part_of_speech != last_part_of_speech) {
+            string_builder.append(part_of_speech_name(definition.part_of_speech));
+            string_builder.append("\n\n");
+
+            last_part_of_speech = definition.part_of_speech;
+        }
+
+        if (n > 1)
+            string_builder.append(AK::String::formatted("  {}. ", sense.sensenum));
+
+        string_builder.append(definition.text);
+        string_builder.append("\n\n");
+    }
+
+    return string_builder.string_view();
+}
+
+void Dictionary::prefix_query(
+    AK::String query,
+    int start_index,
+    int& first_index,
+    int& num_indices) const
+{
+    if (m_num_words == 0) {
+        return;
+    }
+
+    first_index = start_index;
+
+    if (query == "") {
+        num_indices = m_num_words - start_index;
+        return;
+    }
+
+    /* First, find the new first_index by padding the query with zeroes
+           and finding the first entry that's >= it. */
+    char query_chars[17] = {};
+    snprintf(query_chars, 17, "%s", query.characters());
```

Comment:
Is there any reason to use `snprintf` here instead of just `String`?

Reply:
What I'm trying to do here is basically just make sure I have a block of 16 bytes that I can do fast binary comparisons with all the way down the index until I hit the right word.

So, that's why I reached for `char[]`, and once I had a `char[]` and some stuff to put in it, I instinctively reached for `snprintf` — though, if there's a preferred way to do it, I'd be happy to, of course. You might have to spell it out for me though 😄

Reply:
This would probably require to change `m_index` to something more C++-y, this code feels to be more like C. Maybe it would be better to wait for someone more experienced to take a view on it :)

Reply:
`m_index` is just a pointer into `wordnet.bin`.

The index is literally just every word, padded to 16 chars, one after the other, like this:

```
0000  a...............
0010  banana..........
0020  catamaran.......
  [...]
9980  yodel...........
9990  zebra...........
```

It starts at the top, looks at the first letter of each word until it gets a match, then continues from there looking at the second letter, etc.

It's kinda dumb but it works and is fast. I started with the dumbest possible thing and then just made it cleverer until it was fast enough.

(It would be even faster to store all the first letters together, then the second ones, and so on, but I think I've optimized this enough. It's decently fast.)

Reply:
Might be best left for a subsequent PR, but, instead of inventing your own packed binary format, how hard would it be to extend LibSQL to be able to parse and walk the database in its unmodified SQLite-3 format?

Reply:
If you went that route, it would let you use LibSQL internally in the database to query it at runtime, and avoid the awkward pointer math and packed binary format you have here.

Reply:
I don't know, but probably very. To be honest, I was more interested in writing a dictionary app than working on an SQL implementation.

When you say awkward pointer math, do you mean all pointer math is awkward, or the pointer math in this particular PR is awkward?

Reply:
BTW, I don't want to come across like I think my ad-hoc binary thing is better than a good DB layer. Having systems share common parts like that is really nice.

I would, though, probably prefer to defer that to a future PR, as you said.

I hope this can be considered an attempt at a usable, MVP Dictionary app, rather than the best possible Dictionary app.

Reply:
Sure, MVP is a good way to approach this project.

I meant pointer math in general, I didn't actually take that close of a look at this file. Just enough to wonder "wait we started with a database, isn't there a database library in the system that could parse it instead of this manually packed format?"

That being said if code looks like C that's a good sign that there's probably some type safer C++-y way to do it hiding in the wings.

Reply:
Bit more info on the WordNet DB:
* SQLite isn't its original format, the maintainers distribute it in a custom text format and it's someone else that maintains the SQLite distro.
* The WordNet SQLite .db file is 400Mib (100Mib zipped), and I just get the sense that Serenity likes things to be smaller and tighter than that. Of course, you can cut it down a bit, but at that point you're no longer just parsing the file in its original format.

On C-like code, I tend to agree. If I am doing pointer/byte-level stuff I do at least like to wrap it in a nice C++ interface, which is what the `Dictionary` class is intended as.

It would be possible to be a bit more C++-y even there, though, for sure. For instance, whenever I have an offset into the mapped file, I considered having, instead of `int32_t`, an `Offset<T>` which dereferences to a `T`.

I decided against it in the end  — the cost/benefit ratio didn't feel quite right — but I could do it like that, if you like, if it'd improve this PR's chances 🙂

Reply:
I've updated Dictionary.cpp to make it more type-safe. This removes almost all of the ugly casting and pointer arithmetic. I think it's a big improvement, so thanks for pushing me in that direction. What do you think?

Reply:
Thanks for the context on the wordnet database! I'm not 100% sold on the Offset<T> wrapper, but it seems to be getting the job done. I'm not sure how C-like it was before b/c I didn't look that closely but it seems in decentish shape at this point to me.  I left a longer review below.